pkg/auth/mtls_authhandler.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	return &authResponse{
+		expirationTime: *expirationTime,
+	}, nil
+}
+
+func (m *mtlsAuthHandler) authType() policy.AuthType {
+	return policy.AuthTypeMTLSSpiffe
+}
+
+func (m *mtlsAuthHandler) listenForConnections(upstreamCtx context.Context, ready chan<- struct{}) {
+	// set up TCP listener
+
+	ctx, cancel := context.WithCancel(upstreamCtx)
+	defer cancel()
+
+	var lc net.ListenConfig
+	l, err := lc.Listen(ctx, "tcp", fmt.Sprintf(":%d", m.cfg.MTLSListenerPort))
+	if err != nil {
+		m.log.WithError(err).Fatal("Failed to start mTLS listener")
+	}
+	go func() { // shutdown socket goroutine
+		<-ctx.Done()
+		l.Close()
+	}()
+
+	m.log.WithField(logfields.Port, m.cfg.MTLSListenerPort).Info("Started mTLS listener")
+	ready <- struct{}{} // signal to hive that we are ready to accept connections
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			conn, err := l.Accept()
+			if err != nil {
+				m.log.WithError(err).Error("Failed to accept connection")
+				if errors.Is(err, net.ErrClosed) {
+					m.log.Info("mTLS listener socket got closed, re-opening")
+					return
+				}
+				continue
+			}
+			go m.handleConnection(ctx, conn)
+		}
+	}
+}
+
+func (m *mtlsAuthHandler) handleConnection(ctx context.Context, conn net.Conn) {
+	defer conn.Close()
+
+	caBundle, err := m.cert.GetTrustBundle()
+	if err != nil {
+		m.log.WithError(err).Error("failed to get CA bundle")
+		return
+	}
+
+	tlsConn := tls.Server(conn, &tls.Config{
+		ClientAuth:     tls.RequireAndVerifyClientCert,
+		GetCertificate: m.GetCertificateForIncomingConnection,
+		MinVersion:     tls.VersionTLS13,
+		ClientCAs:      caBundle,
+	})
+	defer tlsConn.Close()
+
+	if err := tlsConn.HandshakeContext(ctx); err != nil {
+		m.log.WithError(err).Error("failed to perform TLS handshake")
+	}
+}
+
+func (m *mtlsAuthHandler) GetCertificateForIncomingConnection(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	m.log.WithField("SNI", info.ServerName).Debug("Got new TLS connection")
+	id, err := m.cert.SNIToNumericIdentity(info.ServerName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get identity for SNI %s: %w", info.ServerName, err)
+	}
+
+	return m.cert.GetCertificateForIdentity(id)
+}
+
+func (m *mtlsAuthHandler) onStart(ctx hive.HookContext) error {
+	m.log.Info("Starting mTLS auth handler")
+
+	listenCtx, cancel := context.WithCancel(context.Background())
+	m.cancelSocketListen = cancel
+
+	ready := make(chan struct{})
+	go m.listenForConnections(listenCtx, ready)
+	<-ready // wait for the socket to be ready
+	return nil
+}
+
+func (m *mtlsAuthHandler) onStop(ctx hive.HookContext) error {
+	m.log.Info("Stopping mTLS auth handler")
+	m.cancelSocketListen()
+	return nil
+}
+
+// verifyPeerCertificate is used for Go's TLS library to verify certificates
+func (m *mtlsAuthHandler) verifyPeerCertificate(id *identity.NumericIdentity, caBundle *x509.CertPool, certChains [][]*x509.Certificate) (*time.Time, error) {
```

Comment:
nitpicking: any reason we're returning a pointer to `time.Time` instead of the value?

Reply:
It is to be able to send no time without having to rely on time.zero() as well as to keep the logic of sending and checking the time more simple and cleaner to the reader. I looked into changing it but it would require a bunch more if statements in the code, still happy to consider it if you think it can be problematic